source/common/quic/envoy_quic_stream.h

```diff
@@ -70,31 +70,22 @@ class EnvoyQuicStream : public virtual Http::StreamEncoder,
       }
     }
 
-    if (status_changed && !in_decode_data_callstack_) {
-      // Avoid calling this while decoding data because transient disabling and
-      // enabling reading may trigger another decoding data inside the
-      // callstack which messes up stream state.
-      if (disable) {
-        // Block QUIC stream right away. And if there are queued switching
-        // state callback, update the desired state as well.
-        switchStreamBlockState(true);
-        if (unblock_posted_) {
-          should_block_ = true;
-        }
-      } else {
-        should_block_ = false;
-        if (!unblock_posted_) {
-          // If this is the first time unblocking stream is desired, post a
-          // callback to do it in next loop. This is because unblocking QUIC
-          // stream can lead to immediate upstream encoding.
-          unblock_posted_ = true;
-          connection()->dispatcher().post([this] {
-            unblock_posted_ = false;
-            switchStreamBlockState(should_block_);
-          });
-        }
-      }
+    if (!status_changed || *unblock_posted_) {
+      return;
     }
+    // If this is the first time blocking or unblocking stream is desired, post a
+    // callback to do it in the end of the event loop. This is because unblocking a quic
+    // stream can trigger another decoding inside the callstack which messes up stream state, and
+    // blocking a quic stream in a QUICHE callstack is not expected by QUICHE.
+    *unblock_posted_ = true;
+    connection()->dispatcher().post(
```

Comment:
+1

Please consider use of a SchedulableCallback owned by the stream.  Call scheduleCallbackNextIteration() to schedule the callback; call cancel() or delete the SchedulableCallback to avoid getting called once you have entered the stream termination sequence.

Reply:
Switched to use SchedulableCallback. Instead of scheduleCallbackNextIteration(), I used scheduleCallbackInCurrentIteration() because all we need is to not call stream SetUnblocked() in place.